prov/rxm/src/rxm_cq.c

```diff
@@ -549,6 +549,10 @@ static ssize_t rxm_rndv_handle_wr_data(struct rxm_rx_buf *rx_buf)
 		}
 	}
 
+	/* BUG: This is forcing a state change without knowing what state
+	 * we're currently in.  This loses whether we processed the completion
+	 * for the original send request.
+	 */
 	RXM_UPDATE_STATE(FI_LOG_CQ, tx_buf, RXM_RNDV_WRITE);
```

Comment:
isn't this the original send request completion handler path?

Reply:
This function is handling a receive completion.  That receive carries the RMA write target information.  The problem is that the tx_buf can be in 1 of 2 states here: RXM_RNDV_TX or RXM_RNDV_WRITE_DATA_WAIT.  The first state indicates that the write request was sent, but we have not handled the send completion.  The second state indicates that we processed the send completion.

It would probably be cleaner if the rendezvous (read or write) and atomic code maintained a separate boolean value to indicate if there's a send completion outstanding or not.  The send completion being processed or not is what's blowing up the number of states, which makes following that code path difficult.